lib/jwt/decode.rb

```diff
@@ -49,10 +49,12 @@ def verify_signature
       raise(JWT::VerificationError, 'Signature verification failed')
     end
 
-    def verify_algo
+    def verify_and_set_algos
       raise(JWT::IncorrectAlgorithm, 'An algorithm must be specified') if allowed_algorithms.empty?
       raise(JWT::IncorrectAlgorithm, 'Token is missing alg header') unless alg_in_header
-      raise(JWT::IncorrectAlgorithm, 'Expected a different algorithm') unless valid_alg_in_header?
+
+      @algos = find_valid_algos_in_header
```

Comment:
Setting the `@algos` in the middle of everything feels a little wrong.

We are interested in the intersection between the allowed_algorithms and what is specified in the token header, right?

Could there be a method (maybe a memoizing one) that would be used in the `verify_algos` and `verify_signature_for?`?
for example
```
def allowed_and_valid_algorithms
  @allowed_and_valid_algorithms ||= allowed_algorithms.select { |alg| alg.valid_alg?(alg_in_header) }
end
```